### PR TITLE
fixing calculation of average

### PR DESCRIPTION
**Pull request by grzanka**
This PR introduces couple of changes in how multiple runs are averaged:
- online version of mean and variance calculation methods are introduced and documented in `averaging.py` module
- error attribute of page is now initialized to None, instead of array of same shape as data - this reduces memory if no error calculation is requested
- tests are adjusted accordingly
- support for python 3.8 is dropped to fully benefit from type hinting syntax introduced in 3.9

### CODE REVIEW

**Comment by grzanka** [comment]
To test create a virtual environment and run:

```
pip install git+https://github.com/DataMedSci/pymchelper.git@722-check-if-weighted-average-from-multiple-runs-are-ok#egg=pymchelper
```

If working on HPC (Ares) remember to include new python `ml python` first, but do not load official pymchelper package via `ml pymchelper`.

**Comment by grzanka** [comment]
TODO:

- [ ] weights needs to be added to all aggregators `update` method
- [ ] weights needs to be seeded (as number of simulated primaries) here https://github.com/DataMedSci/pymchelper/blob/d6c5ac3a5f85b5d74381dac325c83a5a3d584c2d/pymchelper/input_output.py#L120

**Comment by grzanka** [comment]
@nbassler that should be ready for review now

File: pymchelper/averaging.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+statistics we sometimes parallelise the simulation of hundreds or thousands of jobs (when using HPC clusters).
+In such case it's not feasible to load all the data into memory and calculate the average in one go, using standard
+functions from numpy library. Instead, we need to calculate the average in an online manner,
+i.e. by updating the state of respective aggregator object with each new binary output file read.
+
+Such approach results in a significant reduction of memory usage and is more numerically stable.
+
+This module contains several classes for aggregating data from multiple files:
+- Aggregator: base class for all other aggregators
+- WeightedStatsAggregator: for calculating weighted (using reliability, not frequency weights) mean and variance
```
Comment (nbassler) [inline pymchelper/averaging.py:19]: what is reliability?

File: pymchelper/averaging.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+- ConcatenatingAggregator: for concatenating data
+- SumAggregator: for calculating sum instead of variance
+- NoAggregator: for cases when no aggregation is required
```
Comment (nbassler) [inline pymchelper/averaging.py:22]: WeightedStatsAggregator  = SH_POSTPROC_AVER 
ConcatenatingAggregator = SH_POSTPROC_APPEND 
SumAggregator = SH_POSTPROC_SUM
NoAggregator = SH_POSTPROC_NONE

how about case SH_POSTPROC_NORM, which would be count/primary ?

File: pymchelper/averaging.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        raise NotImplementedError(f"Update function not implemented for {self.__class__.__name__}")
+
+    def error(self, **kwargs):
+        """Default implementation of error function, returns None."""
+        logging.debug("Error calculation not implemented for %s", self.__class__.__name__)
+
+    @property
+    def updated(self):
+        """Check if the aggregator was updated."""
+        if isinstance(self.data, float) and np.isnan(self.data):
```
Comment (nbassler) [inline pymchelper/averaging.py:62]: is NANs here used to trigger update? (maybe describe the logic in docstring)

File: pymchelper/averaging.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+    def error(self, **kwargs):
+        """Error calculation function, can be used to calculate standard deviation or standard error."""
+        logging.debug("Calculating error with kwargs: %s", kwargs)
+        if 'error_type' in kwargs:
+            if kwargs['error_type'] == 'stddev':
+                return self.stddev
+            if kwargs['error_type'] == 'stderr':
+                return self.stderr
+        return None
```
Comment (nbassler) [inline pymchelper/averaging.py:170]: is error_type always set to either stddev or stderr? (maybe throw exception with clear error message)

File: pymchelper/averaging.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    def variance_sample(self):
+        """
+        Unbiased estimate of the variance.
+        The bias of the weighted estimator if (1 - sum w_i^2 / W_n^2), or in other words:
+        1 - "sum of squares of weights" / "square of sum of weights".
+        For all equal weights the bias is 1 - n * w^2/((n * w)^2) = 1 - 1/n which
+        leads to the well known formula for the sample variance.
+        Here we use the weighted version of the formula.
+        """
+        return self._accumulator_S / (self.total_weight - (self._total_weight_squared / self.total_weight))
```
Comment (nbassler) [inline pymchelper/averaging.py:145]: check if self.total_weight != 0 ?

File: pymchelper/averaging.py (hunk 1 above)
Comment (grzanka) [reply pymchelper/averaging.py:19]: Jargon from statistics, I wasn't familiar with them before as well.

Reliability weights are the weights which do not sum up to the 1. For example w_1=100, w_2 = 3000, w_3 = 2000. This is in contrary to frequency weights like w_1 = 0.1, w_2 = 0.4, w_3 = 0.5.

See for example:
https://en.wikipedia.org/wiki/Weighted_arithmetic_mean

image

File: pymchelper/averaging.py (hunk 2 above)
Comment (grzanka) [reply pymchelper/averaging.py:22]: The code in `averaging.py`  is SHIELD-HIT12A agnostic. It just calculates various types of aggregators. 
The `SH_POSTPROC` numbers are not used in this file, to see how it is used in context of SHIELD-HIT12A take a look at `input_output.py`

`SH_POSTPROC_*` corresponds to https://github.com/DataMedSci/pymchelper/blob/93cfda176bce289504c2a72bb1122166f6e080ac/pymchelper/input_output.py#L22-L29 

also take a look at:
https://github.com/DataMedSci/pymchelper/blob/93cfda176bce289504c2a72bb1122166f6e080ac/pymchelper/input_output.py#L119-L120

the `page_normalized` is a `page` feature extracted from BDO file.

In some cases (for example normalization = 2) the BDO binary file contain number which is not normalized per-primary. Then such normalization happens in the `fromfile` method of the pymchelper reader:
https://github.com/DataMedSci/pymchelper/blob/93cfda176bce289504c2a72bb1122166f6e080ac/pymchelper/readers/shieldhit/reader_bdo2019.py#L165-L171

This way estimators obtained using `fromfile` method is already normalized "per-particle".  That holds for all cases: TLET/DLET and Dose/Fluence and others.

With such assumptions its OK to apply weighted status for `AveragingCumulative` (2) and `AveragingPerPrimary` (3)

That fact needs a comment in the code I think !

File: pymchelper/averaging.py (hunk 5 above)
Comment (grzanka) [reply pymchelper/averaging.py:145]: That could happen only if you haven't updated the aggregator. Good idea to throw a `ValueError` in that case

File: pymchelper/averaging.py (hunk 4 above)
Comment (grzanka) [reply pymchelper/averaging.py:170]: We could have user which requested no error to be calculated. The error type is defined here:
https://github.com/DataMedSci/pymchelper/blob/93cfda176bce289504c2a72bb1122166f6e080ac/pymchelper/estimator.py#L12-L19

if no error is requested then I think we could safely return None. 
Anyway good point - I will improve the code

File: pymchelper/averaging.py (hunk 3 above)
Comment (grzanka) [reply pymchelper/averaging.py:62]: NaN is initial value assigned to data when constructing dataclass object, see:
https://github.com/DataMedSci/pymchelper/blob/93cfda176bce289504c2a72bb1122166f6e080ac/pymchelper/averaging.py#L48

when the object is constructed and update method is not called yet we are in the not-updated-yet state.

anyway good point - I will add more comments

File: pymchelper/averaging.py (hunk 4 above)
Comment (grzanka) [reply pymchelper/averaging.py:189]: Fixed in f82d04e4d5f9772b12485ad29571f06ac2196b9b

Comment (nbassler) [reply pymchelper/averaging.py:192]: second last return None is redundant

File: pymchelper/averaging.py (hunk 2 above)
Comment (nbassler) [reply pymchelper/averaging.py:22]: Just FYI, the SH_POSTPROC_SUM was not supposed to be per primary, even after processing the jobs. It is indeed a total count, I guess useful for debugging

Comment (grzanka) [reply pymchelper/averaging.py:22]: Implemented in set of last 6 commits, see https://github.com/DataMedSci/pymchelper/pull/723/files/93cfda176bce289504c2a72bb1122166f6e080ac..52947dc09b1cc2af172213102541f243ca132095

File: pymchelper/input_output.py [hunk 6]
```diff
@@ -112,6 +112,16 @@ def fromfilelist(input_file_list,
         if not result:
             return None
 
+        # mapping from page-normalisation (from SHIELD-HIT12A) to aggregator (from pymchelper.averaging)
+        # SHIELD-HIT12A uses integers for normalisation, here we stick to corresponding enum values
+        # note that for some normalisation types, the same aggregator is used
+        # these are AveragingCumulative (like dose) and AveragingPerPrimary (like LET)
```
[the remark this answers is not in the document]
Comment (nbassler) [reply pymchelper/input_output.py:118]: should be fixed in SH12A I guess

[the remark this answers is not in the document]
Comment (nbassler) [reply pymchelper/input_output.py:118]: Ok, it is not a mistake, it is correct as you say @grzanka  
dose, fluence are summed in BDO, and thus not reported per particle. Therefore it is sufficient to sum x_i from all jobs and divide by total number of particles for X/primary

File: pymchelper/input_output.py [hunk 7]
```diff
@@ -20,7 +20,15 @@
 
 
 class AggregationType(IntEnum):
-    """Enum for different types of aggregation."""
+    """
+    Enum for different types of aggregation.
+    Below few examples of such aggregation types used in SHIELD-HIT12A:
+      - NoAggregation is used for density (RHO) and material scorer.
```
Comment (nbassler) [inline pymchelper/input_output.py:26]: X = x_0 (values are taken from the first job instance)

File: pymchelper/input_output.py [hunk 8]
```diff
@@ -20,7 +20,15 @@
...
+      - Sum is used for particle counter (COUNT).
```
Comment (nbassler) [inline pymchelper/input_output.py:27]: X = sum^j_i x_i          # j = number of jobs

File: pymchelper/input_output.py [hunk 9]
```diff
@@ -20,7 +20,15 @@
...
+      - AveragingCumulative is used for dose and fluence scorers.
+      - AveragingPerPrimary is used for LET scorers (TLET and DLET).
```
Comment (nbassler) [inline pymchelper/input_output.py:29]: X = (sum^j_i x_i  * n_i) / sum^j_i n_i)    # meaning x_i is already an average so we need a weighted average over the all jobs

**Review by nbassler — COMMENTED** [review]
maybe add the above comments for clarity?

(or maybe a table listing this incl stddev and stderr calculation)

File: pymchelper/input_output.py (hunk 7 above)
Comment (nbassler) [reply pymchelper/input_output.py:26]: just to be clear, it will be taken from the first file in the input

Comment (grzanka) [reply pymchelper/input_output.py:26]: Thats correct, but unfortunately all the files are being read, but their contents being ignored. See
https://github.com/DataMedSci/pymchelper/blob/52947dc09b1cc2af172213102541f243ca132095/pymchelper/averaging.py#L223-L235

File: pymchelper/input_output.py (hunk 8 above)
Comment (grzanka) [reply pymchelper/input_output.py:27]: Thats correct, for implementation:

https://github.com/DataMedSci/pymchelper/blob/52947dc09b1cc2af172213102541f243ca132095/pymchelper/averaging.py#L208-L219

File: pymchelper/input_output.py (hunk 7 above)
Comment (grzanka) [reply pymchelper/input_output.py:26]: I will add some comments

Comment (nbassler) [reply pymchelper/input_output.py:26]: but that is fine!

File: pymchelper/input_output.py [hunk 10]
```diff
@@ -20,7 +20,15 @@
 
 
 class AggregationType(IntEnum):
-    """Enum for different types of aggregation."""
+    """
+    Enum for different types of aggregation.
+    Below few examples of such aggregation types used in SHIELD-HIT12A:
+      - NoAggregation is used for density (RHO) and material scorer.
+      - Sum is used for particle counter (COUNT).
+      - AveragingCumulative is used for dose and fluence scorers.
```
[the remark this answers is not in the document]
Comment (grzanka) [reply pymchelper/input_output.py:28]: The question is what do you mean by `x_i` here ? Data saved in BDO file or data retrieved by `fromfile` function ?

**Review by nbassler — APPROVED** [review]
Checked this against rev49 by sha256sum for [timepix](https://github.com/APTG/2022_DCPT_LET_timepix) data . Only difference found was png images, which have changed due to the /prim.

OK

**Pull request merged**